Add defaults to vDeployments to avoid uninitialized variables

ajtowns opened the pull request:
Adds default values for `vDeployments` in `consensus/params.h` so that undefined behaviour is avoided if a deployment is not initialized. Also adds a check in the unit tests to alert if this is happening, since even if it doesn't invoke undefined behaviour it's probably a mistake.

ajtowns commented:
Something like this was previously proposed in #21401. Deployments have been left uninitialised in the past, see #20157 and also [taproot activation in elements](https://github.com/ElementsProject/elements/pull/1044#pullrequestreview-757081000) so this seems worth having the compiler check for similar mistakes. I think it ends up pretty readable too.

MarcoFalke commented on src/deploymentinfo.cpp:
```diff
@@ -6,7 +6,7 @@
 #include <consensus/params.h>
 
-const struct VBDeploymentInfo VersionBitsDeploymentInfo[Consensus::MAX_VERSION_BITS_DEPLOYMENTS] = {
+const struct std::array<VBDeploymentInfo, Consensus::MAX_VERSION_BITS_DEPLOYMENTS> VersionBitsDeploymentInfo{{
```
```
C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\deploymentinfo.cpp(9,109): error C2220: the following warning is treated as an error [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_common\libbitcoin_common.vcxproj]
C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\deploymentinfo.cpp(9,109): warning C4099: 'std::array<VBDeploymentInfo,2>': type name first seen using 'class' now seen using 'struct' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_common\libbitcoin_common.vcxproj]
C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\deploymentinfo.h(20): message : see declaration of 'std::array<VBDeploymentInfo,2>' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_common\libbitcoin_common.vcxproj]

ajtowns replied:
Huh, I can't get clang to duplicate that error, even with `-Wmismatched-tags`. Anyway, fixed.

ajtowns commented on src/util/syntaxhelp.h:
```diff
@@ -0,0 +1,107 @@
+    {
+        static_assert(E < N, "too many initializers for array");
+        static_assert(E == O, "array initializer element is out of order");
+        if constexpr (E < N) {
```
FYI: this `if constexpr` is only here to avoid redundant warnings that `std::get<N>` and `step<N+1>` would generate.

ajtowns reviewed (commented):
Went over this with @amitiuttarwar and noticed some things.

PastaPastaPasta commented on src/chainparams.cpp:
```diff
@@ -55,6 +56,25 @@ static CBlock CreateGenesisBlock(uint32_t nTime, uint32_t nNonce, uint32_t nBits
+static constexpr util::NamedArgument<int, struct minactheight_arg> _minactheight;
+
+template<size_t O>
+static inline util::ArrayInitElement<O, Consensus::BIP9Deployment> Dep(decltype(_bit)::Arg bit, decltype(_start_time)::Arg start_time, decltype(_timeout)::Arg timeout, decltype(_minactheight)::Arg minactheight)
```
I don't really like the name 'Dep'. Giving this a more descriptive name would, imo, be beneficial

ajtowns replied:
Renamed to `DepSettings`

PastaPastaPasta commented on src/chainparams.cpp:
```diff
@@ -55,6 +56,25 @@ static CBlock CreateGenesisBlock(uint32_t nTime, uint32_t nNonce, uint32_t nBits
+static constexpr util::NamedArgument<int64_t, struct timeout_arg> _timeout;
+static constexpr util::NamedArgument<int, struct minactheight_arg> _minactheight;
+
+template<size_t O>
```
can you use a different name than `O`? I confused it for `0` when first reading

ajtowns replied:
Renamed to `Offset`; also, `I` renamed to `Init`

PastaPastaPasta commented on src/chainparams.cpp:
```diff
@@ -55,6 +56,25 @@ static CBlock CreateGenesisBlock(uint32_t nTime, uint32_t nNonce, uint32_t nBits
+template<size_t O>
+static inline util::ArrayInitElement<O, Consensus::BIP9Deployment> Dep(decltype(_bit)::Arg bit, decltype(_start_time)::Arg start_time, decltype(_timeout)::Arg timeout, decltype(_minactheight)::Arg minactheight)
+{
+    Consensus::BIP9Deployment dep;
+    dep.bit = bit.v;
+    dep.nStartTime = start_time.v;
+    dep.nTimeout = timeout.v;
+    dep.min_activation_height = minactheight.v;
+    return util::ArrayInitElement<O, Consensus::BIP9Deployment>{dep};
```
Can you add a constructor to BIP9Deployment that takes these values? That way this method body collapses to just 1 line, and we'd then be passing a temporary

ajtowns replied:
gcc and clang both optimise them to the same thing, and saying `dep.bit = bit.v` etc seems a bit more obvious that everythings in the right order than `BIP9Deployment(bit.v, start_time.v, timeout.v, minactheight.v)` does. (per [godbolt](https://godbolt.org/z/Th4cc7j9f), VS seems to generate some moderately weird assembly, but changing how the temporary BIP9Deployment struct is created initialised doesn't seem to improve things)

PastaPastaPasta reviewed (commented):
Strong concept ACK, I have a few suggestions / concerns

JeremyRubin commented:
Concept ACK

is there other ongoing work on deployment tech that is motivating this change? Or just good maitenance.

ajtowns commented:
> is there other ongoing work on deployment tech that is motivating this change? Or just good maitenance.

There were a bunch of clean ups that got dropped to get speedy trial done speedily; seems a good time to reconsider them now that versionbits logic isn't actively being used, and before we update to whatever might come after speedy trial. This seemed a good one to start with since it's caused bugs twice now.

PastaPastaPasta commented:
Please see https://github.com/PastaPastaPasta/dash/commit/768a84f8ef1d544efd88ac9fae316df8777328fa to constexpr all the things

ajtowns commented:
Marked as draft, since both #23508 (simple, user-visible improvement) and #23536 (changes how consensus is enforced so annoying to re-test after rebase) should be merged first, in my opinion.

PastaPastaPasta commented on src/versionbits.cpp:
```diff
@@ -194,15 +194,15 @@ class VersionBitsConditionChecker : public AbstractThresholdConditionChecker {
 public:
     explicit VersionBitsConditionChecker(Consensus::DeploymentPos id_) : id(id_) {}
-    uint32_t Mask(const Consensus::Params& params) const { return ((uint32_t)1) << params.vDeployments[id].bit; }
+    uint32_t Mask(const Consensus::Params& params) const { return ((uint32_t)1) << params.vDeployments.at(id).bit; }
```
please convert this to be a c++11 style functional cast

MarcoFalke replied:
I don't think functional casts were added in C++11, pretty sure they existed before already.

PastaPastaPasta replied:
Yeah, seems like c++11 added the `int{}` syntax but not `int()`. Learn smth new everyday

ajtowns commented:
Rebased. Removed that `.at()` change based on https://www.erisian.com.au/bitcoin-core-dev/log-2022-02-02.html#l-301 . Still in draft for #23536

ajtowns committed:
consensus/params: set default values for BIP9Deployment

While chainparams should explicilty set values for each possible
entry in vDeployments, in the past that has been missed resulting
in potential undefined behaviour due to accessing unitinitialized
data. Reduce the severity of future bugs of that nature by providing
benign default values. Adds a unit test to alert if the default value
is not overwritten for the real chains (NEVER_ACTIVE/NEVER_ACTIVE rather
than NEVER_ACTIVE/NO_TIMEOUT).

ajtowns commented:
Removes all the template magic; and downgrades the "you forgot to initialize a deployment" check from compile-time to the unit tests, and doesn't convert to `std::array`. On the upside, this means this PR no longer conflicts with #23536.

Based on [this week's meeting](https://bitcoin.jonasschnelli.ch/ircmeetings/logs/bitcoin-core-dev/2022/bitcoin-core-dev.2022-03-10-19.00.log.html#l-114), also switches to using designated initializers.
